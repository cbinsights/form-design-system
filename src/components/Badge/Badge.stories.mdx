import { Meta, Story, Preview, Props } from "@storybook/addon-docs/blocks";
import { withKnobs, text, optionsKnob as options } from '@storybook/addon-knobs';
import { arrayToOptions, ImportPath } from 'components/util/storybook';

import Badge, { SIZES } from '.';

<Meta title="Components/Badge" decorators={[withKnobs]} component={Badge} />

# Badge

<ImportPath component={Badge} section="media" />

<Story name="Knobs" parameters={{ docs: { disable: true }}}>
  <Badge 
    label={text('label', 'Badge')} 
    size={options('size', arrayToOptions(SIZES), undefined, {
      display: 'inline-radio',
    })}
  />
</Story>

<Preview isExpanded>
  <Badge 
    label="Badge"
  />
</Preview>

<Props of={Badge} />

### Sizes

<Preview>
  <Story name="Sizes">
    {SIZES.map(size => (
      <div className="margin--right--half display--inlineBlock">
        <Badge size={size} label="Badge" />
      </div>
    ))}
  </Story>
</Preview>
