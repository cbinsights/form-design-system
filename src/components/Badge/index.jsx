import React from 'react';
import PropTypes from 'prop-types';
import cx from 'classnames';

export const SIZES = ['xs', 's'];

const Badge = ({ label, size = 's', ...rest }) => (
  <div {...rest} className={cx('fdsBadge', 'wrap--singleLine', `fdsBadge--${size}`)}>
    {label}
  </div>
);

Badge.propTypes = {
  /** Text inside the button */
  label: PropTypes.string.isRequired,
  /** Controls size display */
  size: PropTypes.oneOf(SIZES),
};

export default Badge;
