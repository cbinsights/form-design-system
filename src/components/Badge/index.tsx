import React from 'react';
import cc from 'classcat';
import Icon from 'components/Icon';

export const VALID_TYPES = ['update', 'count'] as const;

export interface BadgeProps {
  /**
   * String to render inside the Badge bubble.
   * The `element` type should be reserved for FDS icons.
   */
  label?: string | Icon | JSX.Element;

  /** When true, an indicator dot will be rendered when the `message` is empty */
  showDot?: boolean;

  /** Type of Badge */
  type?: typeof VALID_TYPES[number];
}

const Badge = ({ label, showDot = false, type = 'update' }: BadgeProps): JSX.Element => {
  const isUpdate = type === 'update';

  return (
    <div className="fds-badgeWrapper alignChild--center--center">
      {(label || showDot) && (
        <div
          aria-label={isUpdate ? 'Notification badge' : 'Count badge'}
          className={cc([
            'fds-badge',
            'padding--x--xs fontSize--xs alignChild--center--center',
            {
              'shape--circle fds-badge--dot color--white': !label,
              'color--white bgColor--red border--none': isUpdate,
              'color--aqua': !isUpdate,
              'bgColor--aqua': !isUpdate && !label,
            },
          ])}
        >
          <span className="fds-badge-label wrap--singleLine--truncate">{label}</span>
        </div>
      )}
    </div>
  );
};

export default Badge;
