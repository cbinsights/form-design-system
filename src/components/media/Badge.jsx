import React from 'react';
import PropTypes from 'prop-types';
import cc from 'classcat';

export const SIZES = ['xs', 's'];

const Badge = ({ label, size = 's', ...rest }) => (
  <div {...rest} className={cc(['fdsBadge', 'wrap--singleLine', `fdsBadge--${size}`])}>
    {label}
  </div>
);

Badge.propTypes = {
  /** Text inside the button */
  label: PropTypes.string.isRequired,
  /** Controls size display */
  size: PropTypes.oneOf(SIZES),
};

export default Badge;
