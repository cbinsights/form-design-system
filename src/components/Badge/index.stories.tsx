import React from 'react';
import { StoryObj } from '@storybook/react';
import Flex from 'components/Flex';
import FlexItem from 'components/FlexItem';
import { StoryWrapper, StoryItem } from 'util/storybook-docs/StoryLayout';
import Badge, { BadgeProps } from '.';

export const Primary: StoryObj<BadgeProps> = {
  args: {
    label: '42',
  },
};

export const Variations: StoryObj<BadgeProps> = {
  render: (args: BadgeProps): JSX.Element => (
    <StoryWrapper>
      <StoryItem>
        <Badge {...args} label="42" />
      </StoryItem>
      <StoryItem>
        <Badge {...args} label="42" type="count" />
      </StoryItem>
      <StoryItem>
        <Badge {...args} />
      </StoryItem>
      <StoryItem>
        <Badge {...args} type="count" />
      </StoryItem>
    </StoryWrapper>
  ),
  parameters: {
    docs: {
      description: {
        story:
          'Badge uses the `update` type by default, which is used to notify users of new updates. The `count` type can be used to show a count of total items. When the `label` is left undefined, both types will collapse into a dot.',
      },
    },
  },
};

export const WithTableRows: StoryObj<BadgeProps> = {
  render: (args: BadgeProps): JSX.Element => (
    <Flex>
      <FlexItem shrink>
        <div
          style={{
            width: '100%',
            height: '32px',
          }}
        >
          <Badge {...args} />
        </div>
      </FlexItem>
      <FlexItem>
        <div
          className="bgColor--haze"
          style={{
            width: '100%',
            height: '32px',
          }}
        />
      </FlexItem>
    </Flex>
  ),
  parameters: {
    docs: {
      description: {
        story:
          'To demonstrate a table row is "new", place the badge to the left of the first column. The bubble will center itself within the nearest parent.',
      },
    },
  },
};

export default {
  component: Badge,
  title: 'components/Badge',
  parameters: {
    componentSubtitle: 'Renders a bubble with an optional label.',
    docs: {
      fds: {
        related: ['CornerPosition'],
      },
      subtitle: 'ok',
      description: {
        component:
          'The bubble will be centered within the parent container you place it in. To render a `Badge` in the top right corner of an element, use the `CornerPosition` component.',
      },
    },
  },
};
