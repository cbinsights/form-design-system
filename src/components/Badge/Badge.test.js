import React from 'react';
import { shallow } from 'enzyme';

import Badge from '.';

describe('Badge component', () => {

  it('matches snapshot (default props)', () => {
    const component = shallow(<Badge label="badge" />);
    expect(component).toMatchSnapshot();
  });

  it('matches snapshot (set all props)', () => {
    const component = shallow(<Badge label="badge" size="xs" data-test="test" />)
    expect(component).toMatchSnapshot();
  });
});
