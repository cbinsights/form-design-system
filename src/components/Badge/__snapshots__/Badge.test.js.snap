// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Badge component matches snapshot (default props) 1`] = `
<div
  className="fdsBadge wrap--singleLine fdsBadge--s"
>
  badge
</div>
`;

exports[`Badge component matches snapshot (set all props) 1`] = `
<div
  className="fdsBadge wrap--singleLine fdsBadge--xs"
  data-test="test"
>
  badge
</div>
`;
